### DISCUSSION TITLE
Using `lower("columnName")` in `where` for case insensitive searches | Postgres + Prisma 2

### DISCUSSION THREAD

**Discussion by pavelsavva** [prisma/prisma#10545] [Q&A]
I apologize if I'm missing something obvious here, but how do I implement something like `select * from table where lower("columnName") = 'value';`? I have manually created an index on `lower("columnName")` via SQL. I've tried
```
where: {
        columnName: {
             equals: 'value',
             mode: 'insensitive'
        }
}
```
however it results in a full database scan. [Prisma documentation](https://www.prisma.io/docs/concepts/components/prisma-client/case-sensitivity) links[ this Postgres page](https://www.postgresql.org/docs/current/indexes-expressional.html) that suggests indexes on `lower()` to achieve case insensitive searches, however, I can't figure out how to use `lower()` in the `where` clause of Prisma find methods.

I'm using Postgres and Prisma 2.

**Comment by austincrim** [comment]
Hey @pavelsavva 👋, thanks for the question.

You can execute this as a raw SQL query using our [`prisma.$queryRaw` API](https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access#queryraw).

> however it results in a full database scan.

Could you talk to this point a little more? Did you see a noticeable performance hit?

**Reply by pavelsavva** [reply]
Hey @austincrim, thank you for your reply!

> You can execute this as a raw SQL query using our [`prisma.$queryRaw` API](https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access#queryraw).

This is the approach I'm using at the moment, but I was wondering if there's a native way to do it with Prisma. We have multiple tables that have indexes on `lower()` fields, so it is becoming pretty complex to maintain all the raw queries. 

> > however it results in a full database scan.
> 
> Could you talk to this point a little more? Did you see a noticeable performance hit?

Yes, I have a database with ~1.1M listings, and using the `where` query mentioned in my first post takes around 2-3 min to complete versus 100-200ms if I'm using `prisma.$queryRaw` API. 

Since the database is indexed on the `lower()` field, using the `where` query mentioned above results in O(n) time complexity vs O(logn) for the `prisma.$queryRaw`. So we are looking at the full scan of 1000000 and vs just 6 rows. 

Thank you!

**Comment by matthewmueller** [comment]
Hey @pavelsavva sorry for the trouble, can you share a reproduction or steps to reproduce with us? I'd like to try this myself and run `EXPLAIN` on the case-insensitive query to see if the index kicks in. There's a chance [the documentation](https://www.prisma.io/docs/concepts/components/prisma-client/case-sensitivity#performance) is wrong or needs to be clarified.

**Comment by supalarry** [comment]
@matthewmueller I think Postgres [documentation](https://www.postgresql.org/docs/current/indexes-expressional.html) explains this:
```
For example, a common way to do case-insensitive comparisons is to use the lower function:

SELECT * FROM test1 WHERE lower(col1) = 'value';

This query can use an index if one has been defined on the result of the lower(col1) function:

CREATE INDEX test1_lower_col1_idx ON test1 (lower(col1));
```

I am running in the same problem as @pavelsavva: we have e-mail column in the table, and if I create an index where e-mail is lowercase, then the WHERE clause has to use LOWER function for the e-mail too.

If LOWER is not used in the search query, then having LOWER e-mail index is worthless.

Finally, @austincrim rawQuery can be used, but then if it's a common search term like e-mail it's not a very scalable solution - each developer would first have to remember to write a rawQuery and also it's tedious to do.

**Reply by pavelsavva** [reply]
Hi @supalarry, we ended up using [`@citext`](https://www.postgresql.org/docs/current/citext.html) type for the case-insensitive columns and all queries worked as expected [without `mode` specified](https://www.prisma.io/docs/concepts/components/prisma-client/case-sensitivity).

**Comment by AviBueno** [comment]
Circling back to this in 2025, I don't see lower case index support in Prisma's [indexes](https://www.prisma.io/docs/orm/prisma-schema/data-model/indexes) documentation.
Is there a plan to support lower case indices?

**Reply by ludralph** [reply]
Hi @AviBueno 

This is not on our radar at the moment but we would get to this at some point. You can read more how we intend to manage issues moving forward [here](https://github.com/prisma/prisma/discussions/25772)

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;

**Comment by RaphaDeveloper** [comment]
I am in the same situation. Any news here?
